File: codeWars/so_many_permutations!/so_many_permutations!.js
// So Many Permutations!

function permutations(str) {
    const results = [];

    // Convert the string to an array for easier manipulation
    const chars = str.split('');

    // Recursive helper function to generate permutations
    function generate(current, remaining) {
        if (remaining.length === 0) {
            results.push(current.join(''));
        } else {
            for (let i = 0; i < remaining.length; i++) {
                const next = current.concat(remaining[i]);
                const remainingChars = [...remaining.slice(0, i), ...remaining.slice(i + 1)];
                generate(next, remainingChars);
            }
        }
    }

    // Start the generation process
    generate([], chars);

    // Remove duplicates using Set and convert back to an array
    const uniqueResults = [...new Set(results)];

    return uniqueResults;
}

console.log(permutations('a')); // Output: ['a']
console.log(permutations('ab')); // Output: ['ab', 'ba']
console.log(permutations('abc')); // Output: ['abc', 'acb', 'bac', 'bca', 'cab', 'cba']
console.log(permutations('aabb')); // Output: ['aabb', 'abab', 'abba', 'baab', 'baba', 'bbaa']

File: codeWars/katas/so_many_permutations!/so_many_permutations!.py
# So Many Permutations!

def permutations(string):
    results = set()

    # Recursive helper function to generate permutations
    def generate(current, remaining):
        if len(remaining) == 0:
            results.add(''.join(current))
        else:
            for i in range(len(remaining)):
                next_permutation = current + [remaining[i]]
                remaining_chars = remaining[:i] + remaining[i+1:]
                generate(next_permutation, remaining_chars)

    # Convert the string to a list of characters for easier manipulation
    chars = list(string)

    # Start the generation process
    generate([], chars)

    # Sort the permutations and return the result
    return sorted(list(results))


print(permutations('a'))     # Output: ['a']
print(permutations('ab'))    # Output: ['ab', 'ba']
print(permutations('abc'))   # Output: ['abc', 'acb', 'bac', 'bca', 'cab', 'cba']
print(permutations('aabb'))  # Output: ['aabb', 'abab', 'abba', 'baab', 'baba', 'bbaa']
